### PR TITLE
fix(ecs): don't resurrect removed required components on insert_if_new

### LINKED ISSUE #24554: commands.insert_if_new inserts Required Components also if not new
Of the two ways of thinking about Required Components [as discussed here](https://github.com/bevyengine/bevy/issues/23393#issuecomment-4071920015), I am using Required Components to "include" other components at insertion, not to constantly require them. This seems to be the intent from [the docs](https://docs.rs/bevy/latest/bevy/prelude/trait.Component.html#required-components) as well:  

> Components can specify Required Components. If some [Component](https://docs.rs/bevy/latest/bevy/prelude/trait.Component.html) A requires [Component](https://docs.rs/bevy/latest/bevy/prelude/trait.Component.html) B, then **when A is inserted**, B will also be initialized and inserted (if it was not manually specified).   

But I've found that the required Component `B` is also inserted when `A` is **not** being inserted. What I did is this:  

* Insert `(A,B)`  as `my_entity`  
* Remove `B` from `my_entity`
* Perform `my_entity_commands.insert_if_new(A)`  
* Observe that, surprisingly, there is now a `B`, although the `A` was not inserted.

## Bevy version and features

Bevy version: `0.18.1`, default features  

```bash
# rustup -V
info: The currently active `rustc` version is `rustc 1.94.0-nightly (31010ca61 2025-12-16)`
```

## What you did

I'll attach the repro project files with a minimal test.  Run `cargo test` to see it fail.

[Repro.zip](https://github.com/user-attachments/files/28688779/Archive.zip)

## What went wrong

I was expecting [insert_if_new](https://docs.rs/bevy/latest/bevy/prelude/struct.EntityCommands.html#method.insert_if_new) to have no effect if the provided Component is _not_ new. Something like "it checks whether the Component exists on that Entity. If yes, it does nothing. If no, it inserts it and triggers hooks and required components appropriately".

What it does instead is it enforces the required Components even if the Component itself was already present. This is a logic bug in my code, where I have a Marker Component that should initially be present (achieved by `#[require(`ing it) and eventually gets removed.  

## Possible Workaround

I suspect I could work around this bug by using Component Hooks to perform the check myself. I did not check, but I would be surprised if the insertion hook would also be called without any insertion happening.  When/if I try this, I will update this post.

## Reproduction Test Case

```rust
mod repro3 {
    use bevy::prelude::*;
    #[derive(Component, Reflect, Default)]
    #[reflect(Component)]
    pub struct RequiredMarkerComponent {}

    #[derive(Reflect, Component, Default)]
    #[reflect(Component)]
    #[require(RequiredMarkerComponent)]
    pub(crate) struct MainComponent {}

    #[test]
    fn simple_insert_if_new_test() {
        // Create an empty world
        let mut world = World::new();
        // Spawn a new entity, with a Component and another, required, Component.
        let mut my_entity = world.spawn(MainComponent::default());

        // Remove the required Component.
        my_entity.remove::<RequiredMarkerComponent>();
        // Be very certain that we did remove the Marker Component
        assert!(my_entity.get::<RequiredMarkerComponent>().is_none());
        // Be very certain that we did not remove the other Component.
        my_entity
            .get::<MainComponent>()
            .expect("This component should still exist...");
        // Add the Component only if new (it's not new. this should do nothing.)
        my_entity.insert_if_new(MainComponent::default());
        match my_entity.get::<RequiredMarkerComponent>() {
            None => {
                // All good.
            }
            Some(_) => panic!(
                "This RequiredMarkerComponent should not exist, because the MainComponent was not new."
            ),
        }
    }
}

```

## Is It a Bug?  
I am of the opinion that this behavior is very unexpected and not as documented, but I see the possibility for the viewpoint that it is working as-intended too... If the consensus is actually that this is fine, I think it should still be documented clearer in the docs of `insert_if_new()` and Required Components.

### PR DESCRIPTION
**Pull request by satyakwok**
Fixes #24554.

`insert_if_new` was re-adding a required component after it had been removed, when the component requiring it was already present — so the call wasn't the no-op it should've been.

The required-components transition is cached on one archetype edge shared by `Replace` and `Keep`, so the target archetype always included the required component (gating just the value init would leave it present-but-uninitialized). I added a separate `Keep` edge that leaves out required components only pulled in by already-present components. Components the insert actually adds still bring their required ones, so mixed bundles like `insert_if_new((A, B))` (A present, B new) still work. `Replace` is unchanged.

Added tests for the repro, the mixed-bundle case, "new component still gets its required", and the `Replace` path.

## Benchmarks

`insert_simple` via criterion `--baseline`, before vs after. This was on a busy shared machine so the numbers are rough:

- `insert_simple/base`: ~688µs → ~593µs (criterion called it -15%, but that's just run-to-run noise here — the change shouldn't actually make inserts faster)
- `insert_simple/unbatched`: ~1.11ms → ~1.13ms ("no change in performance detected")

No regression on the insert hot path — picking the keep/replace edge is a single branch on a cached lookup, and the extra set-building only happens once per edge at cold cache-build, not per insert.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
I was [promised benchmarks](https://github.com/bevyengine/bevy/issues/24554#issuecomment-4684187207) :) Easy to forget; can you please update your PR description to include some rough numbers?

**Comment by satyakwok** [comment]
sorry, those got cut when I trimmed the description down. Added them back.

Short version: `insert_simple/unbatched` shows no change, and `base` measured ~15% faster but that's just noise on my machine (the change shouldn't actually speed inserts up). No regression on the hot path.

File: crates/bevy_ecs/src/bundle/insert.rs [hunk 1]
```diff
@@ -538,17 +544,24 @@ impl BundleInfo {
         components: &Components,
         observers: &Observers,
         archetype_id: ArchetypeId,
+        // When `true`, build the `InsertMode::Keep` edge: required components pulled in only by
+        // components already present in the source archetype are excluded from the target.
+        keep: bool,
```
Comment (chescock) [inline crates/bevy_ecs/src/bundle/insert.rs]: Why use `bool` here instead of `InsertMode`?  The custom enum is a little more self-documenting.

File: crates/bevy_ecs/src/bundle/insert.rs [hunk 2]
```diff
@@ -571,6 +590,11 @@ impl BundleInfo {
 
         for (index, component_id) in self.iter_required_components().enumerate() {
             if !current_archetype.contains(component_id) {
+                // Under `Keep`, skip required components pulled in only by components already
+                // present in the source archetype — they must not be (re-)added (issue #24554).
+                if keep && !required_by_added.contains(&component_id) {
```
Comment (chescock) [inline crates/bevy_ecs/src/bundle/insert.rs]: This `contains` will do an O(N) scan.  Do we need to worry about a quadratic blowup here?

File: crates/bevy_ecs/src/bundle/insert.rs [hunk 3]
```diff
@@ -57,7 +58,11 @@ impl<'w> BundleInserter<'w> {
         archetype_id: ArchetypeId,
         bundle_id: BundleId,
         change_tick: Tick,
+        insert_mode: InsertMode,
     ) -> Self {
+        // `Keep` (`insert_if_new`) uses a distinct archetype edge that excludes required components
+        // pulled in only by already-present components (issue #24554).
```
Comment (chescock) [inline crates/bevy_ecs/src/bundle/insert.rs]: Minor nit, but I don't think we usually include links to resolved issues in comments (here and below).  

```suggestion
        // pulled in only by already-present components.
```

File: crates/bevy_ecs/src/bundle/insert.rs [hunk 4]
```diff
@@ -562,6 +575,12 @@ impl BundleInfo {
                 added.push(component_id);
                 // SAFETY: component_id exists
                 let component_info = unsafe { components.get_info_unchecked(component_id) };
+                if keep {
+                    // This explicit component is newly added, so its (transitive) required
+                    // components are legitimately new and must remain in the `Keep` target.
+                    required_by_added
+                        .extend(component_info.required_components().all.keys().copied());
```
Comment (chescock) [inline crates/bevy_ecs/src/bundle/insert.rs]: I think there is an edge case here if we have a required component with different constructors.  

The Bundle will store the constructor used by the first requiree in the bundle.  But if we want the semantics of `insert_if_new` to be the same as if existing components were absent from the bundle, then we might need a constructor from a later requiree.  

That is, if I modify your test below to have them share a required component: 

```rust
#[derive(Component)]
struct Req(usize);

#[derive(Component)]
#[require(Req(1))]
struct A;

#[derive(Component)]
#[require(Req(2))]
struct B;

let mut world = World::new();
let id = world.spawn(A).id();
world.entity_mut(id).remove::<Req>();
world.entity_mut(id).insert_if_new((A, B));
```

What value should `Req` have?  I would expect the semantics to be "`A` already exists, so this is the same as `insert(B)`, which requires `Req(2)`".  But I think this implementation says "`(A, B)` requires `Req(1)`, and `B` requires `Req`, so we include `Req(1)`".  

If we want the other semantics, then I think you need to accumulate the constructors in an `IndexMap` like in `BundleInfo::new`.

Comment (chescock) [inline crates/bevy_ecs/src/bundle/insert.rs]: Hmm, do we also want the new behavior for recursively required components?  They currently behave more-or-less the same as `insert_if_new`, and a difference might be surprising.  

That is, suppose you have:

```rust
#[derive(Component)]
#[require(B)]
struct A;

#[derive(Component)]
#[require(C)]
struct B;

#[derive(Component)]
struct C;

let mut world = World::new();
let id = world.spawn(B).id();
world.entity_mut(id).remove::<C>();
world.entity_mut(id).insert_if_new(A);
```

Then `insert_if_new(A)` will insert `C` because `A` recursively requires it through `B`, even though `insert_if_new(B)` would not have.  

If we want to change that behavior, too, then I think we need to do the full depth-first search here and give up the `RequiredComponents::all` and `BundleInfo::required_component_constructors` caches completely.  But maybe that's not so bad if we're caching everything on archetype edges anyway?

**Review by chescock — COMMENTED** [review]
The change makes sense, but I'd like to understand and document the intended behavior on the edge cases.  

Oh, and I guess we should document the behavior on the normal cases, too :).  They don't currently say anything about how required components and `insert_if_new` interact.  Maybe add something to the `Component` docs in the "Required Components" section?  Or on `insert_if_new`?  

Do we need a migration guide for this?  It's a behavior change to existing APIs.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 1 above)
Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: I'll thread `InsertMode` through instead of the bool, reads better.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 2 above)
Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: This one actually goes away with the shared-constructor fix below — I'll switch `required_by_added` to a map (required id -> constructor), so the lookup is O(1) instead of the linear scan.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 3 above)
Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: Agreed, I'll drop the issue-link references from the comments here and below.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 4 above)
Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: The current code takes the first requiree's constructor even when that requiree isn't the one being added. The fix ties into the recursive question below: a per-edge DFS would pick the constructor from the added chain for free; if we keep the current non-DFS approach, I'd accumulate constructors in an `IndexMap` from the added requirees. Either way `insert_if_new((A, B))` with `A` already present should end up with `Req(2)`.

Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: I think it has the same root as the constructor case. A per-edge depth-first walk from the explicit components, stopping at components already in the source archetype, would handle both: it only pulls in required components reachable through a newly-added requirer (so `insert_if_new(A)` with `B` present wouldn't add `C`), and it naturally picks the constructor from the added chain.

Since this is built once per archetype edge and cached, the DFS would be cold-path only — the per-insert hot path doesn't change. So I'd lean toward doing the full DFS here and dropping the `all` / `required_component_constructors` shortcuts for this path. Want me to go that way, or keep the current behavior and document the recursive case as a known difference?

**Comment by satyakwok** [comment]
Thanks for the thorough review. Agreed on the docs. I'll write up how required components and `insert_if_new` interact (normal and edge cases), probably in the Required Components section of the `Component` docs, plus a migration guide entry since it's a behavior change. I'll settle the edge cases above first so the docs match the final behavior.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 4 above)
Comment (chescock) [reply crates/bevy_ecs/src/bundle/insert.rs]: > Want me to go that way, or keep the current behavior and document the recursive case as a known difference?

Either seems reasonable!  I wanted to make sure we thought through the edge cases, but I don't have strong opinions on how they should work, and I don't want to cause any unnecessary scope creep.  As long as there's an explanation of why you took the approach you did then I'm happy to approve as a community reviewer!  The final decision is made by a SME, and they'll do a final review once there are two community approvals.  

And if you want more opinions, you're always welcome to hop on the [Bevy Discord](https://discord.gg/bevy)!  The [#ecs-dev](https://discord.com/channels/691052431525675048/749335865876021248) channel is a good place for building consensus on how ECS features like this ought to work.

Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs]: Thanks! I'll go with the smaller change then, keep the current behavior for the recursive case and document it, rather than dropping the caches for a full DFS. That avoids the scope creep and keeps the per-edge cache.

I'll still fix the shared-constructor case (D), since that one's a genuine wrong value rather than a judgment call: I'll accumulate the constructors from the requirees that are actually being added (so `insert_if_new((A, B))` with `A` present ends up with `B`'s `Req`). And I'll write up the reasoning and the recursive-edge behavior in the docs.

**Comment by satyakwok** [comment]
Pushed the changes:

- `keep: bool` -> `InsertMode` through the edge build (A)
- a shared required component now takes the constructor from a requiree that's actually being added, looked up via a map that's also O(1) instead of the linear scan (D + B)
- dropped the resolved-issue links from the comments (C)
- kept the recursive behavior and documented the `insert_if_new` / required-components interaction in the `Component` "Required Components" docs, plus a migration guide for the behavior change (E + docs)

Added a regression test for the shared-constructor case. All `bevy_ecs` tests and doctests pass.

**Review by chescock — APPROVED** [review]
Looks good!

**Comment by satyakwok** [comment]
@alice-i-cecile this got chescock's approval, and the benchmarks + edge-case docs are in now. Since it's tagged X-Contentious, is there anything else you'd want to see before it can move forward?

**Comment by alice-i-cecile** [comment]
This needs a second approval, per our [Contributing Guide](https://bevy.org/learn/contribute/introduction/) :) I don't have the time to do it justice right now, but it's probably worth saying hi and then asking in #ecs-dev on Discord.

File: _release-content/migration-guides/insert_if_new_required_components.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+struct B;
+
+let id = world.spawn(A).id(); // inserts A and its required B
+world.entity_mut(id).remove::<B>();
+world.entity_mut(id).insert_if_new(A);
+// Before: B is present again.
+// Now:    B stays absent, because A was already present so the insert did nothing.
+```
+
+If you relied on the old behavior to restore a required component, insert it explicitly (`insert(B)`), or use `insert` (`InsertMode::Replace`) instead of `insert_if_new`.
```
Comment (laundmo) [inline _release-content/migration-guides/insert_if_new_required_components.md]: This would not actually replicate the behaviour, as in, ut does re-insert required components, but it also replaces the inserted components value, no?
To replicate it, the `insert` has to happen with the current value as queried.

Also: is InsertMode exposed to the APIs users normally use? The way this is phrased makes it kinda seem like the normal EntityCommands::insert has a new InsertMode param. Should probably clarify which API is talked about here, or leave out/clarify the mention of InsertMode if its not relevant to users.

File: crates/bevy_ecs/src/bundle/insert.rs [hunk 6]
```diff
@@ -538,17 +542,27 @@ impl BundleInfo {
         components: &Components,
         observers: &Observers,
         archetype_id: ArchetypeId,
+        insert_mode: InsertMode,
     ) -> (ArchetypeId, bool) {
-        if let Some(archetype_after_insert_id) = archetypes[archetype_id]
+        // `Keep` (`insert_if_new`) builds a distinct edge that excludes required components pulled
```
Comment (laundmo) [inline crates/bevy_ecs/src/bundle/insert.rs:547]: I dislike that InsertMode has this assumption/behavior attached. It would be nicer if the checking of the insert mode was done clearly before the actual logic for inserting, and didn't need to be passed through so deeply.

I'm not requesting a change, per se, just mentioning something i noticed.

**Review by laundmo — CHANGES_REQUESTED** [review]
The migration guide improvements are part of the requested changes.

My thoughts about InsertMode::Keep semantics are not.

File: _release-content/migration-guides/insert_if_new_required_components.md (hunk 5 above)
Comment (satyakwok) [reply _release-content/migration-guides/insert_if_new_required_components.md]: The  `insert` isn't equivalent since it overwrites the requirer's value too. To get the required component back without changing the requirer, just insert it directly (`insert(B)`). I'll fix the wording.

On InsertMode: it's internal, users don't pass it to insert/insert_if_new, so I'll drop it and reference the user-facing API instead.

Comment (laundmo) [reply _release-content/migration-guides/insert_if_new_required_components.md]: Inserting it directly is not equivalent either, it doesn't work if i take the requirer as a generic from a user, for example

Comment (satyakwok) [reply _release-content/migration-guides/insert_if_new_required_components.md]: got it, i'll update the migration guide to say there is no direct equivalent for the old behavior in the generic case, and avoid suggesting `insert` or `Replace` as workarounds there.

Comment (satyakwok) [reply _release-content/migration-guides/insert_if_new_required_components.md]: Updated in `b2b3bff`. The guide now states there is no direct equivalent in the general case, and explains why: you can't name the required components when the requirer is only known generically, and re-inserting the requirer with `Replace` is not a substitute since it overwrites the requirer's own value. Manually re-inserting a required component is only mentioned for the case where you know its concrete type. Also dropped the `InsertMode::Keep` framing in favor of the `insert_if_new` method names.

File: crates/bevy_ecs/src/bundle/insert.rs (hunk 6 above)
Comment (satyakwok) [reply crates/bevy_ecs/src/bundle/insert.rs:547]: the mode ends up in there because `Keep` and `Replace` can resolve to different target archetypes (`Keep` drops the required components that only come from already-present components), and that's decided and cached where the edge is built. kept this PR scoped to the fix rather than reworking that plumbing. can pull the mode handling into a clearer step in a follow-up if you'd prefer.
